docs/core/dependency-loading/understanding-assemblyloadcontext.md

```diff
@@ -0,0 +1,81 @@
+
+The <xref:System.Runtime.Loader.AssemblyLoadContext> class is unique to .NET Core. This article attempts to supplement the API docs <xref:System.Runtime.Loader.AssemblyLoadContext> with conceptual information.
+
+This article is relevant to developers implementing dynamic loading, especially dynamic loading framework developers.
+
+## What is the <xref:System.Runtime.Loader.AssemblyLoadContext>?
+
+Every .NET Core application implicitly uses the <xref:System.Runtime.Loader.AssemblyLoadContext>.
+It's the runtime's provider for locating and loading dependencies. Whenever a dependency is loaded an <xref:System.Runtime.Loader.AssemblyLoadContext> instance is invoked to locate it.
+
+- It provides a service of locating, loading, and caching managed assemblies and other dependencies.
+
+## What is special about the <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> instance?
+
+The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> instance is automatically populated by the runtime at startup.  It uses the [trusted platform assemblies](trusted-platform-assemblies.md) to locate and find all static dependencies.
+
+It solves the most common dependency loading scenarios.
+
+## How does <xref:System.Runtime.Loader.AssemblyLoadContext> support dynamic dependencies?
+
+<xref:System.Runtime.Loader.AssemblyLoadContext> has various events and virtual functions that can be overridden.
+
+The <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> instance only supports overriding the events.
+
+The articles
+[Managed assembly loading algorithm](loading-managed.md),
+[Satellite assembly loading algorithm](loading-resources.md), and
+[Unmanaged (native) library loading algorithm](loading-unmanaged.md) refer to all the available events and virtual functions.  The articles show each event and functions relative position in the loading algorithms. This article won't reproduce that information.
+
+This section will cover the general principles for the relevant events and functions.
+
+- **Be repeatable**. It's required that a query for a specific dependency will always result in the same response. The same loaded dependency instance must be returned. This requirement is fundamental  for cache consistency.
+  - The exception is that the initial `null` response could be followed by a later non-`null` response.
+  - For managed assemblies in particular, we're creating a <xref:System.Reflection.AssemblyName.Name?displayProperty=nameWithType> to <xref:System.Reflection.Assembly> cache. There can be exactly one <xref:System.Reflection.AssemblyName.Name?displayProperty=nameWithType> in the cache.
+- **Typically don't throw**.  It's expected that these functions return `null` rather than throw when unable to find the requested dependency. Throwing should be restricted to unexpected errors like a corrupt assembly or out of memory.
```

Comment:
I think we should add that "throwing will stop the assembly resolution process and the exception will be propagated to the caller immediately" (Maybe we actually even cache it, but I'm not 100% on that).

Reply:
I'm not sure exactly what happens when we throw either. From what I saw it doesn't propagate to the user code.

Reply:
I just tried with custom ALC and `Resolving` event handler. If the event handler throws we wrap it in `FileLoadException` and throw from the binding attempt. So it is propagated (wrapped).
We don't seem to cache the failure - when called twice, the handler is also called twice.

Same for overriding `Load`.

Interestingly enough we don't even cache successes - we still call Load/Resolving again.

Reply:
Hmmm.  
- I definitely was not seeing the exceptions when working on one of the unit tests in coreclr.  Maybe it was the `*.GetType` tests
- Seems like not caching is a bug or a performance opportunity.

Reply:
I am going to continue to write this up as if the caching is correctly implemented.